### PR TITLE
[LLVM] Add C API support for handling global object metadata.

### PR DESCRIPTION
**Pull request by maleadt**
Upstreaming vendored C API fixes from Julia/LLVM.jl.

### CODE REVIEW

File: llvm/lib/IR/Core.cpp [hunk 1]
```diff
@@ -1062,15 +1062,27 @@ void LLVMReplaceAllUsesWith(LLVMValueRef OldVal, LLVMValueRef NewVal) {
   unwrap(OldVal)->replaceAllUsesWith(unwrap(NewVal));
 }
 
-int LLVMHasMetadata(LLVMValueRef Inst) {
-  return unwrap<Instruction>(Inst)->hasMetadata();
+int LLVMHasMetadata(LLVMValueRef Ref) {
+  Value *Val = unwrap<Value>(Ref);
+  if (auto I = dyn_cast<Instruction>(Val))
```
Comment (nikic) [inline llvm/lib/IR/Core.cpp:1067]:
```suggestion
  if (auto *I = dyn_cast<Instruction>(Val))
```
etc.

File: llvm/lib/IR/Core.cpp [hunk 2]
```diff
@@ -1062,15 +1062,27 @@ void LLVMReplaceAllUsesWith(LLVMValueRef OldVal, LLVMValueRef NewVal) {
...
+    return I->hasMetadata();
+  else if (auto GO = dyn_cast<GlobalObject>(Val))
+    return GO->hasMetadata();
+  else
+    assert(0 && "Expected an instruction or a global object");
```
Comment (nikic) [inline llvm/lib/IR/Core.cpp:1072]:
```suggestion
    llvm_unreachable("Expected an instruction or a global object");
```

File: llvm/lib/IR/Core.cpp [hunk 3]
```diff
@@ -1062,15 +1062,27 @@ void LLVMReplaceAllUsesWith(LLVMValueRef OldVal, LLVMValueRef NewVal) {
   unwrap(OldVal)->replaceAllUsesWith(unwrap(NewVal));
 }
 
-int LLVMHasMetadata(LLVMValueRef Inst) {
-  return unwrap<Instruction>(Inst)->hasMetadata();
+int LLVMHasMetadata(LLVMValueRef Ref) {
```
Comment (nikic) [inline llvm/lib/IR/Core.cpp:1065]: Comments in the header need updating.

File: llvm/lib/IR/Core.cpp [hunk 4]
```diff
@@ -1088,10 +1100,16 @@ static MDNode *extractMDNode(MetadataAsValue *MAV) {
   return MDNode::get(MAV->getContext(), MD);
 }
 
-void LLVMSetMetadata(LLVMValueRef Inst, unsigned KindID, LLVMValueRef Val) {
-  MDNode *N = Val ? extractMDNode(unwrap<MetadataAsValue>(Val)) : nullptr;
+void LLVMSetMetadata(LLVMValueRef Ref, unsigned KindID, LLVMValueRef MAV) {
```
[the remark this answers is not in the document]
Comment (maleadt) [reply llvm/lib/IR/Core.cpp:1103]: Ah, I somehow didn't know about that API. It's a shame that the Value metadata API is (partially) covered by two slightly different C APIs... I'll check if I can do what I need with the `LLVMGlobal` API.

**Comment by maleadt** [comment]
The existing `LLVMGlobal` APIs, which I didn't know about, suffice for my need (https://github.com/maleadt/LLVM.jl/pull/446) so we can close this right now.

**Pull request closed without merging**